haystack_experimental/core/pipeline/pipeline.py

```diff
@@ -229,45 +303,382 @@ def run(  # noqa: PLR0915, PLR0912
+            except PipelineBreakException as e:
+                # Add the current pipeline results to the exception
+                e.results = pipeline_outputs
+                raise
 
-                component_pipeline_outputs = self._write_component_outputs(
-                    component_name=component_name,
-                    component_outputs=component_outputs,
-                    inputs=inputs,
-                    receivers=cached_receivers[component_name],
-                    include_outputs_from=include_outputs_from,
-                )
+            return pipeline_outputs
 
-                if component_pipeline_outputs:
-                    pipeline_outputs[component_name] = deepcopy(component_pipeline_outputs)
-                if self._is_queue_stale(priority_queue):
-                    priority_queue = self._fill_queue(ordered_component_names, inputs, component_visits)
+    def _validate_breakpoints(self, breakpoints: Set[Tuple[str, Optional[int]]]) -> Set[Tuple[str, int]]:
+        """
+        Validates the breakpoints passed to the pipeline.
 
-            return pipeline_outputs
+        Make sure they are all valid components registered in the pipeline,
+        If the visit is not given, it is assumed to be 0, it will break on the first visit.
+        If a negative number is given it means it will break on all visits, e.g.: a component running in a loop.
+
+        :param breakpoints: Set of tuples of component names and visit counts at which the pipeline should stop.
+        :returns:
+            Set of valid breakpoints.
+        """
+
+        processed_breakpoints: Set[Tuple[str, int]] = set()
+
+        for break_point in breakpoints:
+            if break_point[0] not in self.graph.nodes:
+                raise ValueError(f"Breakpoint {break_point} is not a registered component in the pipeline")
+            if break_point[1] is None:
+                break_point = (break_point[0], 0)
+            processed_breakpoints.add(break_point)  # type: ignore  # at this point break_point is Tuple[str, int]
+
+        return processed_breakpoints
+
+    def _check_breakpoints(self, breakpoints: Set[Tuple[str, int]], component_name, instance, component_visits, inputs):
+        """
+        Check if the `component_name` is in the breakpoints and if it should break.
+
+        :param breakpoints: Set of tuples of component names and visit counts at which the pipeline should stop.
+        :param component_name: Name of the component to check.
+        :param component_visits: The number of times the component has been visited.
+        :param inputs: The inputs to the pipeline.
+        :raises PipelineBreakException: When a breakpoint is triggered, with component state information.
+        """
+        matching_breakpoints = [bp for bp in breakpoints if bp[0] == component_name]
+        for bp in matching_breakpoints:
+            visit_count = bp[1]
+            # break at every visit if visit_count is -1
```

Comment:
Let's remove this feature because we will raise an Exception when we break for the first time. Users can stop at every visit by resuming the pipeline run and increasing the `visit_count` in `breakpoints={("component_name", visit_count)},`